daemons/based/based_transaction.c

```diff
@@ -0,0 +1,343 @@
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+based_init_transaction(const pcmk__client_t *client)
+{
+    CRM_ASSERT(client != NULL);
+
+    // A client can have at most one transaction at a time
+    if (get_transaction(client) != NULL) {
+        return pcmk_rc_already;
+    }
+
+    if (transactions == NULL) {
+        transactions = pcmk__strkey_table(NULL, free_transaction);
+    }
+    g_hash_table_insert(transactions, client->id, g_queue_new());
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Add a new CIB request to an existing transaction
+ *
+ * \param[in] client      Client whose transaction to extend
+ * \param[in] request     CIB request
+ * \param[in] privileged  If \p true, the request has write privileges
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+based_extend_transaction(pcmk__client_t *client, xmlNodePtr request,
+                         bool privileged)
+{
+    struct request_data *data = NULL;
+    GQueue *transaction = NULL;
+
+    CRM_ASSERT((client != NULL) && (request != NULL));
+
+    transaction = get_transaction(client);
+    if (transaction == NULL) {
+        return pcmk_rc_no_transaction;
+    }
+
+    data = create_request_data(client, request, privileged);
+    if (data == NULL) {
+        return ENOMEM;
+    }
+
+    g_queue_push_tail(transaction, data);
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Free a CIB client's transaction (if any) and its requests
+ *
+ * \param[in] client  Client whose transaction to discard
+ */
+void
+based_discard_transaction(const pcmk__client_t *client)
+{
+    CRM_ASSERT(client != NULL);
+
+    if (transactions != NULL) {
+        g_hash_table_remove(transactions, client->id);
+    }
+}
+
+/*!
+ * \internal
+ * \brief Validate that a CIB request is supported in a transaction
+ *
+ * \param[in] request  CIB request
+ *
+ * \return Standard Pacemaker return code
+ */
+static int
+validate_transaction_request(const xmlNode *request)
+{
+    const char *host = crm_element_value(request, F_CIB_HOST);
+    const char *op = crm_element_value(request, F_CIB_OPERATION);
+
+    if (!pcmk__str_eq(host, OUR_NODENAME, pcmk__str_null_matches)) {
+        crm_err("Operation targeting another node (%s) is not supported in CIB "
+                "transaction",
+                host);
+        return EOPNOTSUPP;
+    }
+
+    if (!pcmk__str_any_of(op,
+                          CRM_OP_PING,
+                          PCMK__CIB_REQUEST_APPLY_PATCH,
+                          PCMK__CIB_REQUEST_BUMP,
+                          PCMK__CIB_REQUEST_CREATE,
+                          PCMK__CIB_REQUEST_DELETE,
+                          PCMK__CIB_REQUEST_ERASE,
+                          PCMK__CIB_REQUEST_MODIFY,
+                          PCMK__CIB_REQUEST_NOOP,
+                          PCMK__CIB_REQUEST_QUERY,
+                          PCMK__CIB_REQUEST_REPLACE,
+                          PCMK__CIB_REQUEST_UPGRADE,
+                          NULL)) {
```

Comment:
This should be a new boolean in cib_operation_t (actually the booleans should be flags, and there are other op checks that could be flags ...)

Reply:
Makes sense. I'll add a commit to convert to flags and then another to add a new flag for transaction support.